proposals/eni.md

```diff
@@ -0,0 +1,128 @@
+             172.31.0.226 |   +-----------+     |               |       |             +-|--|container  |
+                       +----+ |           |     |ecs-eni bridge |       |             | |  +-----------+
+                       |eth0| | ECS Agent |     |               |       | docker0 bridge|
+                       +----+ +-----------+     |               |       |             | |
+                          |                     |               |       |             | |
+                          |                     |  169.254.x.1  |       |             | |
+                          |                     +---------------+       +-------------+ |
+                          |                                                             |
+                          +-------------------------------------------------------------+
+
+In this diagram, the primary network interface is still being used on the instance
+by the `docker0` bridge to let containers launched with `net=bridge` to communicate
+with entities outside the instance. A secondary network interface has also been
+attached for the purpose of being used by a task (`eth1`). This interface has been
+moved to the `ENI-1` network namespace. All containers of the task will share this
+network namespace.
+
+Tasks (containers in tasks) launched into these namespaces will be addressable
+by the primary IP Address of the respective secondary network interfaces. For
+example, all containers of the task using ENI `ENI-1` will be addressable within
+the VPC using the IP Address `172.31.13.14/20`.
+
+The `ecs-eni-br` bridge is used by the containers to communicate with the
+primary network interface `eth0` for credentials and metadata requests. The
+`veth` pairs between the `ENI-1 namespace` and the `ecs-eni-br` bridge are used
+for this purpose (`ve-en1` and `ve_br+en1`).
+
+ECS Agent will start a `pause` container for performing this configuration. This
+container will only be used to start a process that waits for a signal to
+terminate by invoking the `pause()` POSIX API. ECS Agent then moves the ENI to
+the network namespace of this container using a CNI plugin. It will then launch
+containers of the task with `--net=container:pause-container-id`, thus ensuring
+that they use the `pause` container's network namespace.
+
+#### CNI Plugin Sequence
+1. Plugin to assign ENI to a network namespace:
+	1. Get MAC Address for the ENI from EC2 Instance Metaddata Service
+	1. Get ENI device name on default namespace
+	1. Get n/w gateway mask
+	1. Get Primary IP Address of the ENI
+	1. Move the ENI to container's namespace
+	1. Assign the primary IP Address to interface
+	1. Setup route to internet via the gateway
+	1. Delete entries from the routing table if needed for the ENI device
+	1. Start `dhclient` to renew leases
+
+1. Plugin to establish route to ECS Agent for Credentials
+	1. ECS Agent determines an available local IP Address from `169.254.76.0/24`
```

Comment:
Yeah. It might not be as narrow as `/24` though. We might decide to go with `/22` or `/23` at least. The IP range has not yet been decided. I'll modify this to be a bit more generic to reflect this.